library/psa_crypto_storage.h

```diff
@@ -59,7 +59,11 @@ extern "C" {
  * This limitation will probably become moot when we implement client
  * separation for key storage.
  */
+#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C)
+#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000ffffffff
+#else
 #define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000
+#endif
```

Comment:
Do we need to change `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` based on the presence of ITS or could we always use `0xffff0000ffffffff`?

Reply:
We should only change it for ITS. If we change it always, the following check will fail when compiling in a non-SPM system:
https://github.com/itayzafrir/mbed-crypto/blob/919b89ec656c0693631547cf635774b8cd1ac0fc/library/psa_crypto_slot_management.c#L205-L219
Actually clang failed this because the condition `if( id >= PSA_MAX_PERSISTENT_KEY_IDENTIFIER )` would never be true because id is 32 bits in a non-SPM system. In an SPM supported system, `psa_key_id_t` is 32 bits on the client and a 64 bit type on the service side.

Reply:
Could it be 64-bits always?

Reply:
Can what be 64 bits? If you're asking about the `psa_key_id_t` then not really, it would then make ITS to be 96 bits and we'll be in the same loop again.

Reply:
@Patater I see you point, I overlooked the fact that ITS is a also used in a non-segmented system.

> In a segmented system:
ITS would be present on SPE, so MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C would be used so we can use it. ITS would also be present on NSPE, but we wouldn't define ITS_C for the Mbed Crypto client code?

Not exactly, as the crypto client code is implemented just as a proxy, I think it's meaningless for the crypto client side whether ITS is enabled or not. Same thing goes for `PSA_MAX_PERSISTENT_KEY_IDENTIFIER`, it's not used by the client code and it isn't part of the PSA Crypto API.

> In a non-segmented system:
The server and client Mbed Crypto are the same. ITS is available, but we'd not define MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C?

I guess we can't really do that...

---

So, since `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` is not part of the official PSA Crypto API it might be meaningless to other PSA Crypto implementations so I'm not sure if placing it in `crypto_types.h` makes sense...IMO it would make more sense placing it `crypto_extra.h` and keeping different copies of `crypto_extra.h` in Mbed OS, one for client and for server, same as is already being done for `crypto_structs.h` and is planned (https://github.com/ARMmbed/mbed-os/pull/9575) to be done with `crypto_types.h`. However, this forces us to maintain yet another file which is different for client/server. I think that a simpler solution is to just change (919b89ec656c0693631547cf635774b8cd1ac0fc):

From:
```c
#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C)
#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000ffffffff
#else
#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000	
#endif
```
To:
```c
#if defined(MBEDTLS_PSA_CRYPTO_SPM)
#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000ffffffff
#else
#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff0000	
#endif
```

`MBEDTLS_PSA_CRYPTO_SPM` is set only when compiling the _server side_, see https://github.com/ARMmbed/mbed-os/blob/57d8915a899a59999d7e2a9695e20a8a18cb54bd/targets/targets.json#L7905

Please let me know if this is an acceptable solution and I will update this PR.

Reply:
I changed it, also changed the `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` value from 0xffff0000ffffffff to 0xffff000000000000 when compiled for server side. I amended the commit, old branch https://github.com/itayzafrir/mbed-crypto/tree/ipc-support-64-bit-key-ids-2.

Reply:
If a user calls `psa_create_key(PSA_KEY_LIFETIME_PERSISTENT, 0xFFFFFF52, ...)`, how do we know they won't ever get a persistent allocation with `PSA_CRYPTO_ITS_RANDOM_SEED_UID` as the storage UID for their key? `0xFFFFFF52` is certainly less than `PSA_MAX_PERSISTENT_KEY_IDENTIFIER`. Don't we need to explicitly block this UID when we select the storage ID? Previously, we depended on the `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` checks to block allocation of this storage location.

Reply:
There's no need for an explicit block on `PSA_CRYPTO_ITS_RANDOM_SEED_UID`.

In a system where SPM is present and as of https://github.com/ARMmbed/mbed-os/pull/9575, if a user were to call `psa_open_key(PSA_KEY_LIFETIME_PERSISTENT, PSA_CRYPTO_ITS_RANDOM_SEED_UID, &handle)` or `psa_create_key(PSA_KEY_LIFETIME_PERSISTENT, PSA_CRYPTO_ITS_RANDOM_SEED_UID, &handle)` the ID (`psa_key_id_t`) sent from the client (`PSA_CRYPTO_ITS_RANDOM_SEED_UID` in this case or any other ID for that matter) is translated by the server (crypto partition) into a different ID prior to calling the actual Mbed Crypto library. 

The ID (`psa_key_id_t`) on the server side is 64 bits and it translated from what the client sent (32 bits) in the following way:
Place the 32 bit ID received from the client in the upper 32 bits of the 64 bit `psa_key_id_t`, and the calling partition id (32 bits) in the lower 32 bits of the 64 bit `psa_key_id_t`.

| 32 bits PSA_CRYPTO_ITS_RANDOM_SEED_UID | 32 bits CALLING PARTITION ID |
|---|---|

The table shows what the actual ID sent to ITS will looks like.

Example: If a client were to call `psa_open_key` or `psa_create_key` with `PSA_CRYPTO_ITS_RANDOM_SEED_UID` as the ID, the actual ID value sent to ITS will be:
1. If calling from NSPE (partition id `-1`) - ID sent to ITS = `0xffffff52ffffffff`
2. If calling from different partition (for example partition id `3`) - ID sent to ITS = `0xffffff5200000003`

So there's no chance of overwriting the injected `PSA_CRYPTO_ITS_RANDOM_SEED_UID`.
Moreover, the results from the examples above would not pass the check against `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` as both  `0xffffff52ffffffff` and `0xffffff5200000003` are greater than `PSA_MAX_PERSISTENT_KEY_IDENTIFIER`.

All this magic is planned to happen in https://github.com/ARMmbed/mbed-os/pull/9575, this PR is about the required setup for 9575. You might want to take a look at it to see the complete picture.

Reply:
I understand for client/server architectures that this would work. Thanks for your explanation.

How about for K64F or other non-segmented systems?

Reply:
For non-segmented systems this has no effect and everything remains as it is now. `MBEDTLS_PSA_CRYPTO_SPM` is currently only defined for the server side of psoc6 (`FUTURE_SEQUANA_M0_PSA`). See https://github.com/ARMmbed/mbed-os/blob/57d8915a899a59999d7e2a9695e20a8a18cb54bd/targets/targets.json#L7905